Cannot install 2 versions of the same public package

**Describe the bug**
I try installing a different version of a package that was already cached by verdaccio, and it fails

**To Reproduce**
Install version `x.0.0` of a package and then try installing version `y.0.0`.
```
npm set registry http://localhost:4873/

npm i @angular/cdk@7.0.3  # successful, updates package.json to only have version `7.0.3`.

npm i @angular/cdk@9.2.4  # fails, because version is unknown to verdaccio.
```

**Expected behavior**
Both packages are installed successfully.

**Screenshots**
```
npm ERR! code ETARGET
npm ERR! notarget No matching version found for @angular/common@9.1.12.
npm ERR! notarget In most cases you or one of your dependencies are requesting
npm ERR! notarget a package version that doesn't exist.

npm ERR! A complete log of this run can be found in:
npm ERR!     C:\Users\rony.fragin\AppData\Roaming\npm-cache\_logs\2020-09-18T22_16_05_892Z-debug.log
```

**Configuration File (cat ~/.config/verdaccio/config.yaml)**

```
#
# This is the default config file. It allows all users to do anything,
# so don't use it on production systems.
#
# Look here for more config file examples:
# https://github.com/verdaccio/verdaccio/tree/master/conf
#

# path to a directory with all packages
storage: /root/.local/share/verdaccio/storage

store:
  offline-storage:

# path to a directory with plugins to include
plugins: ./plugins

web:
  title: Verdaccio
  # comment out to disable gravatar support
  # gravatar: false
  # by default packages are ordercer ascendant (asc|desc)
  # sort_packages: asc
  # convert your UI to the dark side
  # darkMode: true

listen: 0.0.0.0:4873

# translate your registry, api i18n not available yet
# i18n:
# list of the available translations https://github.com/verdaccio/ui/tree/master/i18n/translations
#   web: en-US

auth:
  htpasswd:
    file: ./htpasswd
    # Maximum amount of users allowed to register, defaults to "+inf".
    # You can set this to -1 to disable registration.
    # max_users: 1000

# a list of other known repositories we can talk to
uplinks:
  npmjs:
    url: https://registry.npmjs.org/
    timeout: 100ms
    cache: true
    fail_timeout: 1m
    max_fails: 100

packages:
  '@*/*':
    # scoped packages
    access: $all
    publish: $authenticated
    unpublish: $authenticated
    proxy: npmjs

  '**':
    # allow all users (including non-authenticated users) to read and
    # publish all packages
    #
    # you can specify usernames/groupnames (depending on your auth plugin)
    # and three keywords: "$all", "$anonymous", "$authenticated"
    access: $all

    # allow all known users to publish/publish packages
    # (anyone can register by default, remember?)
    publish: $authenticated
    unpublish: $authenticated

    # if package is not available locally, proxy requests to 'npmjs' registry
    proxy: npmjs

# You can specify HTTP/1.1 server keep alive timeout in seconds for incoming connections.
# A value of 0 makes the http server behave similarly to Node.js versions prior to 8.0.0, which did not have a keep-alive timeout.
# WORKAROUND: Through given configuration you can workaround following issue https://github.com/verdaccio/verdaccio/issues/301. Set to 0 in case 60 is not enough.
server:
  keepAliveTimeout: 180

middlewares:
  audit:
    enabled: true

# log settings
logs:
  - { type: stdout, format: pretty, level: http }
  #- {type: file, path: verdaccio.log, level: info}
#experiments:
#  # support for npm token command
#  token: false
#  # support for the new v1 search endpoint, functional by incomplete read more on ticket 1732
#  search: false

# This affect the web and api (not developed yet)
#i18n:
#web: en-US
```

**Environment information**
```
*** WARNING: Verdaccio doesn't need superuser privileges. Don't run it under root! ***

Environment Info:

  System:
    OS: Linux 4.15 Ubuntu 18.04.4 LTS (Bionic Beaver)
    CPU: (2) x64 Intel(R) Xeon(R) CPU E5-2680 v2 @ 2.80GHz
  Binaries:
    Node: 10.22.0 - ~/.nvm/versions/node/v10.22.0/bin/node
    npm: 6.14.6 - ~/.nvm/versions/node/v10.22.0/bin/npm
  npmGlobalPackages:
    verdaccio: 4.8.1
```

**Workaround**
I manually go to the url
`https://registry.npmjs.org/@angular/common`
copy the data to the `package.json` in the storage folder and then it works

I ran the following python script to do it for all packages:
```
import json
import urllib

my_list = [<list of package names>]
BASE_DIR = '/root/.local/share/verdaccio/storage'

for item in my_list:
  data = json.loads(urllib.request.urlopen(f"https://registry.npmjs.org/{item}").read())
  with open(f'{BASE_DIR}/{item}/package.json', 'w') as outfile:
    json.dump(data, outfile)
  print(f"done - {item}")
```




every time I install a new version, I need to fetch the `package.json` again from the npm repo... how do I make verdaccio not update `package.json` to only contain specific versions?

--- a/src/OfflinePackageStorage.js
+++ b/src/OfflinePackageStorage.js
@@ -16,6 +16,10 @@ import LocalFS from '@verdaccio/local-storage/lib/local-fs';
  * @see https://github.com/verdaccio/monorepo/tree/master/plugins/local-storage
  */
 export default class OfflinePackageStorage extends LocalFS {
+  constructor(path, logger, offlineMode) {
+    super(path, logger);
+    this.offlineMode = offlineMode;
+  }
   /**
    * Computes a package's definition that only lists the locally available versions.
    *
@@ -43,7 +47,7 @@ export default class OfflinePackageStorage extends LocalFS {
               '[verdaccio-offline-storage/readPackage/readdir] error discovering package "@{packageName}" files: @{err}'
             );
             cb(err);
-          } else {
+          } else if (this.offlineMode) {
             const localVersions = items
               .filter(item => item.endsWith('.tgz'))
               .map(item => item.substring(basename(name).length + 1, item.length - 4));
@@ -52,7 +56,7 @@ export default class OfflinePackageStorage extends LocalFS {
                 packageName: name,
                 count: localVersions.length,
               },
-              '[verdaccio-offline-storage/readPackage/readdir] discovered @{count} items for package: @{packageName}'
+              '[verdaccio-offline-storage/readPackage/readdir] offline mode - discovered @{count} items for package: @{packageName}'
             );
             const allVersions = Object.keys(data.versions);
             const originalVersionCount = allVersions.length;
@@ -61,7 +65,7 @@ export default class OfflinePackageStorage extends LocalFS {
                 packageName: name,
                 count: originalVersionCount,
               },
-              '[verdaccio-offline-storage/readPackage/readdir] analyzing @{count} declared versions for package: @{packageName}'
+              '[verdaccio-offline-storage/readPackage/readdir] offline mode - analyzing @{count} declared versions for package: @{packageName}'
             );
             for (const version of allVersions) {
               if (!localVersions.includes(version)) {
@@ -71,7 +75,7 @@ export default class OfflinePackageStorage extends LocalFS {
                     packageName: name,
                     version,
                   },
-                  '[verdaccio-offline-storage/readPackage/readdir] removed @{packageName}@@{version}'
+                  '[verdaccio-offline-storage/readPackage/readdir] offline mode - removed @{packageName}@@{version}'
                 );
               }
             }
@@ -80,7 +84,7 @@ export default class OfflinePackageStorage extends LocalFS {
                 packageName: name,
                 count: originalVersionCount - Object.keys(data.versions).length,
               },
-              '[verdaccio-offline-storage/readPackage/readdir] removed @{count} versions for package: @{packageName}'
+              '[verdaccio-offline-storage/readPackage/readdir] offline mode - removed @{count} versions for package: @{packageName}'
             );
             data['dist-tags'].latest = Object.keys(data.versions).sort((a, b) => cmp(b, a))[0];
             this.logger.trace(
@@ -88,7 +92,15 @@ export default class OfflinePackageStorage extends LocalFS {
                 packageName: name,
                 latest: data['dist-tags'].latest,
               },
-              '[verdaccio-offline-storage/readPackage/readdir] set latest version to @{latest} for package: @{packageName}'
+              '[verdaccio-offline-storage/readPackage/readdir] offline mode - set latest version to @{latest} for package: @{packageName}'
+            );
+            cb(null, data);
+          } else {
+            this.logger.trace(
+              {
+                packageName: name,
+              },
+              '[verdaccio-offline-storage/readPackage/readdir] online mode - local package @{packageName} found'
             );
             cb(null, data);
           }

--- a/src/OfflineStoragePlugin.js
+++ b/src/OfflineStoragePlugin.js
@@ -16,7 +16,6 @@ import OfflinePackageStorage from './OfflinePackageStorage';
 export default class OfflineStoragePlugin extends LocalDatabase {
   constructor(config, options) {
     super(config, options.logger);
-    this.offlineMode = config.uplinks === undefined || Object.keys(config.uplinks).length === 0;
   }
 
   /**
@@ -89,6 +88,8 @@ export default class OfflineStoragePlugin extends LocalDatabase {
    * @see https://verdaccio.org/docs/en/plugin-storage#api
    */
   getPackageStorage(packageName) {
-    return new OfflinePackageStorage(join(this.config.storage, packageName), this.logger, this.offlineMode);
+    const offlineMode = this.config.uplinks === undefined || Object.keys(this.config.uplinks).length === 0;
+
+    return new OfflinePackageStorage(join(this.config.storage, packageName), this.logger, offlineMode);
   }
 }

--- a/src/OfflineStoragePlugin.js
+++ b/src/OfflineStoragePlugin.js
@@ -16,6 +16,7 @@ import OfflinePackageStorage from './OfflinePackageStorage';
 export default class OfflineStoragePlugin extends LocalDatabase {
   constructor(config, options) {
     super(config, options.logger);
+    this.offlineMode = config.uplinks === undefined || Object.keys(config.uplinks).length === 0;
   }
 
   /**
@@ -88,6 +89,6 @@ export default class OfflineStoragePlugin extends LocalDatabase {
    * @see https://verdaccio.org/docs/en/plugin-storage#api
    */
   getPackageStorage(packageName) {
-    return new OfflinePackageStorage(join(this.config.storage, packageName), this.logger);
+    return new OfflinePackageStorage(join(this.config.storage, packageName), this.logger, this.offlineMode);
   }
 }

